synapse/storage/data_stores/main/state.py

```diff
@@ -363,24 +364,46 @@ def _background_remove_left_rooms_txn(txn):
             if not room_ids:
                 return True, set()
 
+            # exclude rooms where we have active members
             sql = """
                 SELECT room_id
-                FROM current_state_events
+                FROM local_current_membership
                 WHERE
                     room_id > ? AND room_id <= ?
-                    AND type = 'm.room.member'
                     AND membership = 'join'
-                    AND state_key LIKE ?
                 GROUP BY room_id
             """
 
-            txn.execute(sql, (last_room_id, room_ids[-1], "%:" + self.server_name))
-
+            txn.execute(sql, (last_room_id, room_ids[-1]))
             joined_room_ids = {row[0] for row in txn}
+            to_delete = set(room_ids) - joined_room_ids
+
+            # exclude rooms which we are in the process of constructing; these otherwise
+            # qualify as "rooms with no local users", and would have their
+            # forward extremities cleaned up.
+
+            # the following query will return a list of rooms which have forward
+            # extremities that are *not* also the create event in the room.
+
+            sql = """
+                SELECT DISTINCT efe.room_id
+                FROM event_forward_extremities efe
+                LEFT JOIN current_state_events cse ON
+                    cse.event_id = efe.event_id
+                    AND cse.type = 'm.room.create'
+                    AND cse.state_key = ''
+                WHERE
+                    cse.event_id IS NULL
+                    AND efe.room_id > ? AND efe.room_id <= ?
+            """
+
+            txn.execute(sql, (last_room_id, room_ids[-1]))
+            creating_rooms = to_delete.difference(row[0] for row in txn)
+            logger.info("skipping rooms which are being deleted: %s", creating_rooms)
 
-            left_rooms = set(room_ids) - joined_room_ids
+            to_delete.difference_update(creating_rooms)
 
-            logger.info("Deleting current state left rooms: %r", left_rooms)
+            logger.info("Deleting current state left rooms: %r", to_delete)
```

Comment:
Just to walk through the logic here:

1. The returned rooms are ones that have forward extremities that are not the create event (i.e. they are rooms that are not being created currently).
2. `to_delete` is modified to remove the above, meaning we're left with the rooms that *are* being created (saved as `creating_rooms`).
3. `to_delete` is modified to remove `creating_rooms`, leaving only the rooms which are not being currently created. (I think this is equivalent to the intersection of the result of step 1 and `to_delete` at the time of step 1, but that doesn't give you something nice to log, unfortunately.)

Not sure if it matters much, but I find the set operations easier to read using operators, e.g. `to_delete -= creating_rooms` (although this only works if they're both sets, I think).

Reply:
yes, your logic is correct.

I'll add some comments to clarify.

> Not sure if it matters much, but I find the set operations easier to read using operators, e.g. to_delete -= creating_rooms (although this only works if they're both sets, I think).

agreed, and the fact that some of these aren't sets is the reason I had to use the wordy version.

Reply:
The added comments look great! 👍